File: backend/core/ledger.py
"""دفتر حسابات الأطراف — §٧.١ — النظير الخادمي لـ packages/domain/src/ledger.ts."""

from __future__ import annotations

from collections.abc import Iterable
from dataclasses import dataclass
from typing import Literal

from core.money import DomainError, assert_int64

AccountRole = Literal["customer", "supplier"]
Direction = Literal["debit", "credit"]
BalanceStatus = Literal["owes_us", "we_owe", "settled"]


@dataclass(frozen=True)
class LedgerEntry:
    direction: Direction
    amount_minor: int


@dataclass(frozen=True)
class AccountKey:
    party_id: str
    role: AccountRole
    currency: str

    def __str__(self) -> str:
        return f"{self.party_id}|{self.role}|{self.currency}"


def balance_minor(entries: Iterable[LedgerEntry]) -> int:
    balance = 0
    for e in entries:
        if e.amount_minor <= 0:
            raise DomainError("invalid_ledger_amount", str(e.amount_minor))
        assert_int64(e.amount_minor, "amount_minor")
        delta = e.amount_minor if e.direction == "debit" else -e.amount_minor
        balance = assert_int64(balance + delta, "balance")
    return balance


def balance_status(balance: int) -> BalanceStatus:
    if balance > 0:
        return "owes_us"
    if balance < 0:
        return "we_owe"
    return "settled"


def balances_by_account(entries: Iterable[tuple[AccountKey, LedgerEntry]]) -> dict[str, int]:
    groups: dict[str, list[LedgerEntry]] = {}
    for key, entry in entries:
        groups.setdefault(str(key), []).append(entry)
    return {k: balance_minor(v) for k, v in groups.items()}

File: packages/domain/src/ledger.ts
/**
 * دفتر حسابات الأطراف — §٧.١: لا جدول Debt مستقل؛ الرصيد = مجموع المدين − مجموع الدائن.
 * `amount_minor > 0` دائماً والاتجاه debit|credit. دور العميل ودور المورد حسابان منفصلان حتى للشخص
 * نفسه؛ لا مقاصة تلقائية بينهما (المقاصة عملية صريحة بحركتين متقابلتين).
 */
import { DomainError, assertInt64 } from "./money";

export type AccountRole = "customer" | "supplier";
export type Direction = "debit" | "credit";

export interface AccountKey {
  readonly partyId: string;
  readonly role: AccountRole;
  readonly currency: string;
}

export interface LedgerEntry {
  readonly direction: Direction;
  /** > 0 دائماً؛ قيد صفري لا يُنشأ لإرضاء قالب عام. */
  readonly amountMinor: bigint;
}

/** رصيد مفتاح واحد (party, role, currency). موجب العميل «عليه»، سالب المورد «له علينا». */
export function balanceMinor(entries: readonly LedgerEntry[]): bigint {
  let balance = 0n;
  for (const e of entries) {
    if (e.amountMinor <= 0n)
      throw new DomainError("invalid_ledger_amount", e.amountMinor.toString());
    assertInt64(e.amountMinor, "amount_minor");
    balance = assertInt64(
      e.direction === "debit" ? balance + e.amountMinor : balance - e.amountMinor,
      "balance",
    );
  }
  return balance;
}

export type BalanceStatus = "owes_us" | "we_owe" | "settled";

/** تسمية الواجهة «عليه/له» لا تبدّل معنى الاتجاه بحسب الشاشة. */
export function balanceStatus(balance: bigint): BalanceStatus {
  if (balance > 0n) return "owes_us";
  if (balance < 0n) return "we_owe";
  return "settled";
}

export const accountKeyString = (k: AccountKey): string => `${k.partyId}|${k.role}|${k.currency}`;

/** يجمّع القيود بمفتاح الحساب الكامل — لا يجمع العملات ولا الدورين ضمناً (§٦.٤). */
export function balancesByAccount(
  entries: readonly (LedgerEntry & { readonly account: AccountKey })[],
): Map<string, bigint> {
  const groups = new Map<string, LedgerEntry[]>();
  for (const e of entries) {
    const key = accountKeyString(e.account);
    const list = groups.get(key) ?? [];
    list.push(e);
    groups.set(key, list);
  }
  return new Map([...groups].map(([k, list]) => [k, balanceMinor(list)]));
}
